File: kotlin/src/main/kotlin/boj/3000/3151.kt
package boj.`3000`

import java.util.StringTokenizer

/*
https://www.acmicpc.net/problem/3151
3151.합이 0
풀이1.5036ms
 */
fun main() {
    val n = readln().toInt()
    val numbers = mutableListOf<Int>()

    with(StringTokenizer(readln())) {
        while (hasMoreTokens())
            numbers.add(nextToken().toInt())
    }

    numbers.sort()
    var answer = 0L

    fun binarySearchUpperBound(low: Int, high: Int, number: Int): Int {
        var low = low
        var high = high

        while (low + 1 < high) {
            val mid = (low + high) / 2
            if (numbers[mid] <= number) {
                low = mid
            } else {
                high = mid
            }
        }

        return high
    }

    fun binarySearchLowerBound(low: Int, high: Int, number: Int): Int {
        var low = low
        var high = high

        while (low + 1 < high) {
            val mid = (low + high) / 2
            if (numbers[mid] < number) {
                low = mid
            } else {
                high = mid
            }
        }

        return high
    }

    for (i in 0 until n - 2) {
        for (j in i + 1 until n - 1) {
            val target = -(numbers[i] + numbers[j])
            val upperBound = binarySearchUpperBound(j, n, target)
            val lowerBound = binarySearchLowerBound(j, n, target)
            answer += upperBound - lowerBound
        }
    }

    println(answer)
}

File: python/BOJ/3XXX/3151.py
"""
https://www.acmicpc.net/problem/3151
3151.합이 0
풀이1.
"""
import sys
input = sys.stdin.readline


def binary_search_upper_bound(numbers, number):
    low = -1
    high = len(numbers)

    while low + 1 < high:
        mid = (low + high) // 2
        if numbers[mid] <= number:
            low = mid
        else:
            high = mid

    return high


def binary_search_lower_bound(numbers, number):
    low = -1
    high = len(numbers)

    while low + 1 < high:
        mid = (low + high) // 2
        if numbers[mid] < number:
            low = mid
        else:
            high = mid

    return high


n = int(input())
numbers = sorted(list(map(int, input().split())))
answer = 0

for i in range(n - 2):
    for j in range(i + 1, n - 1):
        target = -(numbers[i] + numbers[j])
        upper_bound = binary_search_upper_bound(numbers[j + 1:], target)
        lower_bound = binary_search_lower_bound(numbers[j + 1:], target)
        answer += upper_bound - lower_bound

print(answer)